### PR TITLE
feature/feedback2

### PR DESCRIPTION
**Pull request by AzamatMuratbekov**

### CODE REVIEW

File: src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+    public RatingResponce rating() {
+        RatingResponce ratingResponce = new RatingResponce();
+        ratingResponce.setOne(feedbackRepository.countFeedbackByProductEvaluation((byte) 1));
+        ratingResponce.setTwo(feedbackRepository.countFeedbackByProductEvaluation((byte) 2));
+        ratingResponce.setThree(feedbackRepository.countFeedbackByProductEvaluation((byte) 3));
+        ratingResponce.setFour(feedbackRepository.countFeedbackByProductEvaluation((byte) 4));
+        ratingResponce.setFive(feedbackRepository.countFeedbackByProductEvaluation((byte) 5));
+        ratingResponce.setRating(feedbackRepository.rating());
+        ratingResponce.setAllFeedback(feedbackRepository.allfeedback());
```
Comment (Aidai0404) [inline src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java:126]: AllFeedback is Long type, для чего она? чтобы вывести ID продуктов или лист продуктов?

File: src/main/java/us/peaksoft/gadgetarium/repository/FeedbackRepository.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package us.peaksoft.gadgetarium.repository;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+import us.peaksoft.gadgetarium.entity.Feedback;
+
+@Repository
+public interface FeedbackRepository extends JpaRepository<Feedback, Long> {
+    Integer countFeedbackByProductEvaluation(byte productEvaluation);
```
Comment (Aidai0404) [inline src/main/java/us/peaksoft/gadgetarium/repository/FeedbackRepository.java:10]: считает кол-во отзывов?

**Review by Aidai0404 — COMMENTED** [review]
viewed

File: src/main/java/us/peaksoft/gadgetarium/repository/FeedbackRepository.java (hunk 2 above)
Comment (AzamatMuratbekov) [reply src/main/java/us/peaksoft/gadgetarium/repository/FeedbackRepository.java:10]: нет, оценка продукта например сколько людей поставили 5 звезд

File: src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java (hunk 1 above)
Comment (AzamatMuratbekov) [reply src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java:126]: Общее количество отзывов на продукт

File: src/main/java/us/peaksoft/gadgetarium/controller/FeedbackController.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import us.peaksoft.gadgetarium.dto.FeedbackRequest;
+import us.peaksoft.gadgetarium.dto.FeedbackResponce;
+import us.peaksoft.gadgetarium.dto.RatingResponce;
+import us.peaksoft.gadgetarium.dto.SimpleResponse;
+import us.peaksoft.gadgetarium.service.FeedbackService;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("api/feedback")
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/controller/FeedbackController.java:14]: feedbacks

File: src/main/java/us/peaksoft/gadgetarium/dto/FeedbackResponce.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    private Long id;
+    private String product;
+    private String user;
+    private String firstName;
+    private String lastName;
+    private String feedback;
+    private String adminReplay;
+    private String media;
+    private byte productEvaluation;
+    private Date createdDate;
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/dto/FeedbackResponce.java:24]: Pls change to LocalDate

File: src/main/java/us/peaksoft/gadgetarium/entity/Feedback.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    private Product product;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    private User user;
+
+    private String feedback;
+    private String adminReplay;
+    private String media;
+    private byte productEvaluation;
+    private Date createdDate;
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/entity/Feedback.java:32]: Pls change to LocalDate

Comment (i-Coderr) [reply src/main/java/us/peaksoft/gadgetarium/entity/Feedback.java:32]: and you change to createdAt

File: src/main/java/us/peaksoft/gadgetarium/entity/User.java [hunk 6]
```diff
@@ -65,6 +65,8 @@ public class User implements UserDetails {
     private Payments payments;
     @OneToMany(cascade = CascadeType.ALL, mappedBy = "user")
     private List<Order> orders;
+    @OneToMany(cascade = CascadeType.ALL, mappedBy = "user")
```
Comment (i-Coderr) [inline src/main/java/us/peaksoft/gadgetarium/entity/User.java:68]: Next line pls

**Review by i-Coderr — COMMENTED** [review]
Commented and resolve conversations

File: src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java (hunk 1 above)
Comment (Aidai0404) [reply src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java:126]: аа, все поняла

File: src/main/java/us/peaksoft/gadgetarium/dto/FeedbackResponce.java (hunk 4 above)
Comment (AzamatMuratbekov) [reply src/main/java/us/peaksoft/gadgetarium/dto/FeedbackResponce.java:24]: LocalDate don't has a time, only date

File: src/main/java/us/peaksoft/gadgetarium/entity/Feedback.java (hunk 5 above)
Comment (AzamatMuratbekov) [reply src/main/java/us/peaksoft/gadgetarium/entity/Feedback.java]: done

File: src/main/java/us/peaksoft/gadgetarium/entity/Feedback.java [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+@AllArgsConstructor
+@Entity
+@Table(name = "feedbacks")
+public class Feedback {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    private Product product;
```
Comment (Aidai0404) [inline src/main/java/us/peaksoft/gadgetarium/entity/Feedback.java:23]: you should add Transient variable for Product's id

**Review by Aidai0404 — COMMENTED** [review]
did you test in postman?

**Review by Aidai0404 — COMMENTED** [review]
please also add values for feedback in data.sql

File: src/main/java/us/peaksoft/gadgetarium/dto/FeedbackRequest.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+import lombok.NoArgsConstructor;
+import lombok.Setter;
+
+@Getter
+@Setter
+@NoArgsConstructor
+@AllArgsConstructor
+public class FeedbackRequest {
+    private Long product;
+    private Long user;
```
Comment (Aidai0404) [inline src/main/java/us/peaksoft/gadgetarium/dto/FeedbackRequest.java:14]: write userId instead user

**Review by Aidai0404 — COMMENTED** [review]
also please add URL of feedback's controller in SecurityConfig! And you didn't change the name of your requestMappimg to api/feedbacks

File: src/main/java/us/peaksoft/gadgetarium/controller/FeedbackController.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    public FeedbackResponce update(@PathVariable("id") Long id, @RequestBody FeedbackRequest feedbackRequest) {
+        return feedbackService.update(id, feedbackRequest);
+    }
+
+    @DeleteMapping("{id}")
+    public SimpleResponse delete(@PathVariable("id") Long id) {
+        return feedbackService.delete(id);
+    }
+
+    @PutMapping("/admin/{id}")
```
Comment (Aidai0404) [inline src/main/java/us/peaksoft/gadgetarium/controller/FeedbackController.java:44]: change to PostMapping

File: src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    }
+
+    public FeedbackResponce mapToResponse(Feedback feedback) {
+        FeedbackResponce feedbackResponce = new FeedbackResponce();
+        feedbackResponce.setId(feedback.getId());
+        feedbackResponce.setFeedback(feedback.getFeedback());
+        feedbackResponce.setProductEvaluation(feedback.getProductEvaluation());
+        feedbackResponce.setMedia(feedback.getMedia());
+        feedbackResponce.setAdminReplay(feedback.getAdminReplay());
+        feedbackResponce.setProduct(feedback.getProduct().getName());
```
Comment (Aidai0404) [inline src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java:100]: where also userId(in your request it is as just "user"

**Review by Aidai0404 — COMMENTED** [review]
I've checked how it works. And u've done it correctly by logic. I've tested in PostMan. I've left some comments. It works well. Nice and well done!

**Comment by Aidai0404** [comment]
viewed

File: src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+        simpleResponse.setMessage("Ваш отзыв успешно отправлен");
+        return simpleResponse;
+    }
+
+    @Override
+    public FeedbackResponce update(Long id, FeedbackRequest feedbackRequest) {
+        Feedback feedback1 = feedbackRepository.findById(id).get();
+        feedback1.setFeedback(feedbackRequest.getFeedback());
+        feedback1.setMedia(feedbackRequest.getMedia());
+        feedback1.setCreatedAt(new Date());
```
[the remark this answers is not in the document]
Comment (AzamatMuratbekov) [reply src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java:53]: There are only date, I need time too.Ok I use LocalDateTime

File: src/main/java/us/peaksoft/gadgetarium/controller/FeedbackController.java (hunk 9 above)
Comment (AzamatMuratbekov) [reply src/main/java/us/peaksoft/gadgetarium/controller/FeedbackController.java:44]: Why? If I want to change the answer?

File: src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java (hunk 10 above)
Comment (AzamatMuratbekov) [reply src/main/java/us/peaksoft/gadgetarium/service/impl/FeedbackServiceImpl.java:100]: deleted

**Review by i-Coderr — APPROVED** [review]
Alright

**Pull request merged**